src/h3lib/lib/algos.c

```diff
@@ -677,36 +673,162 @@ void H3_EXPORT(polyfill)(const GeoPolygon* geoPolygon, int res, H3Index* out) {
     BBox* bboxes = malloc((geoPolygon->numHoles + 1) * sizeof(BBox));
     assert(bboxes != NULL);
     bboxesFromGeoPolygon(geoPolygon, bboxes);
-    int minK = bboxHexRadius(&bboxes[0], res);
-    int numHexagons = H3_EXPORT(maxKringSize)(minK);
-
-    // Get the center hex
-    GeoCoord center;
-    bboxCenter(&bboxes[0], &center);
-    H3Index centerH3 = H3_EXPORT(geoToH3)(&center, res);
-
-    // From here on it works differently, first we get all potential
-    // hexagons inserted into the available memory
-    H3_EXPORT(kRing)(centerH3, minK, out);
-
-    // Next we iterate through each hexagon, and test its center point to see if
-    // it's contained in the GeoJSON-like struct
-    for (int i = 0; i < numHexagons; i++) {
-        // Skip records that are already zeroed
-        if (out[i] == 0) {
-            continue;
+
+    // Get the estimated number of hexagons and allocate some temporary memory
+    // for the hexagons
+    int numHexagons = H3_EXPORT(maxPolyfillSize)(geoPolygon, res);
+    // This algorithm assumes that the number of vertices is usually less than
+    // the number of hexagons, but when it's wrong, this will keep it from
+    // failing
+    const Geofence geofence = geoPolygon->geofence;
+    int numVerts = geofence.numVerts;
+    for (int i = 0; i < geoPolygon->numHoles; i++) {
+        numVerts += geoPolygon->holes[i].numVerts;
+    }
+    if (numHexagons < numVerts) numHexagons = numVerts;
+    H3Index* search = calloc(numHexagons, sizeof(H3Index));
+    H3Index* found = calloc(numHexagons, sizeof(H3Index));
```

Comment:
No, because I'm swapping `found` and `search` each major "step" of this algorithm so I know what hexagons to check the next time through the loop.

It also doesn't use 3x the prior memory because the `maxPolyfillSize` equation reduces the number of hexagons dramatically. :)